<a href="https://colab.research.google.com/github/shubhanalytics/PGPDSBA-Projects/blob/main/Proj.01%20-%20Python%20for%20Data%20Analysis/Uber%20Drive.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

UBER DRIVE ANALYSIS : The project is based on the trips made by Uber drivers. Different aspects of the trip are analyzed by using different functions in Python.

Load the necessary libraries. Import and load the dataset with a name uber.

In [1]:
import numpy as np, pandas as pd, seaborn as sns, os
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
print(os.getcwd()) # checking the current working directory and printing in the output to get confirmed

D:\My Data\Professional\Analytics\PGPDSBA-Projects\Proj.01 - Python for Data Analysis


In [3]:
# load the csv file uberdrive.csv
uber = pd.read_csv(r'D:\My Data\Professional\Analytics\PGPDSBA-Projects\Proj.01 - Python for Data Analysis\Uber_Drive.csv') 

uber.head() #read the file and check the first 5 rows

,START_DATE*,END_DATE*,CATEGORY*,START*,STOP*,MILES*,PURPOSE*
0,1/1/2016 21:11,1/1/2016 21:17,Business,Fort Pierce,Fort Pierce,5.1,Meal/Entertain
1,1/2/2016 1:25,1/2/2016 1:37,Business,Fort Pierce,Fort Pierce,5.0,NaN
2,1/2/2016 20:25,1/2/2016 20:38,Business,Fort Pierce,Fort Pierce,4.8,Errand/Supplies
3,1/5/2016 17:31,1/5/2016 17:45,Business,Fort Pierce,Fort Pierce,4.7,Meeting
4,1/6/2016 14:42,1/6/2016 15:49,Business,Fort Pierce,West Palm Beach,63.7,Customer Visit


In [4]:
uber.tail(10)   # Displaying the last 10 records of the dataset

,START_DATE*,END_DATE*,CATEGORY*,START*,STOP*,MILES*,PURPOSE*
1145,12/30/2016 10:15,12/30/2016 10:33,Business,Karachi,Karachi,2.8,Errand/Supplies
1146,12/30/2016 11:31,12/30/2016 11:56,Business,Karachi,Karachi,2.9,Errand/Supplies
1147,12/30/2016 15:41,12/30/2016 16:03,Business,Karachi,Karachi,4.6,Errand/Supplies
1148,12/30/2016 16:45,12/30/2016 17:08,Business,Karachi,Karachi,4.6,Meeting
1149,12/30/2016 23:06,12/30/2016 23:10,Business,Karachi,Karachi,0.8,Customer Visit
1150,12/31/2016 1:07,12/31/2016 1:14,Business,Karachi,Karachi,0.7,Meeting
1151,12/31/2016 13:24,12/31/2016 13:42,Business,Karachi,Unknown Location,3.9,Temporary Site
1152,12/31/2016 15:03,12/31/2016 15:38,Business,Unknown Location,Unknown Location,16.2,Meeting
1153,12/31/2016 21:32,12/31/2016 21:50,Business,Katunayake,Gampaha,6.4,Temporary Site
1154,12/31/2016 22:08,12/31/2016 23:51,Business,Gampaha,Ilukwatta,48.2,Temporary Site


In [5]:
uber.shape  # showing the shape or dimension(number of rows, number of columns) of the dataset

(1155, 7)

In [6]:
uber.size   # showing the size(rows * columns) of the dataset

8085

In [7]:
uber.info()    # displaying the information about all the variables of the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1155 entries, 0 to 1154
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   START_DATE*  1155 non-null   object 
 1   END_DATE*    1155 non-null   object 
 2   CATEGORY*    1155 non-null   object 
 3   START*       1155 non-null   object 
 4   STOP*        1155 non-null   object 
 5   MILES*       1155 non-null   float64
 6   PURPOSE*     653 non-null    object 
dtypes: float64(1), object(6)
memory usage: 63.3+ KB


In the above dataset info, below are the few observations:

1. There are 1155 non-null values in first 6 columns, but there are only 653 non-null values in the 'purpose' column, which means 502 values are null in 'purpose' column. We need to review and correct it.

2. Only the 'Miles' column is in float datatype, rest all columns are in object datatype, which requires our manual review and correction.


In [9]:
uber.isnull().any()    # checking the missing values in the dataset in the boolean(True or False) format

START_DATE*    False
END_DATE*      False
CATEGORY*      False
START*         False
STOP*          False
MILES*         False
PURPOSE*        True
dtype: bool

In [11]:
uber.isnull().sum()  # checking the missing values in the dataset in the integer format

START_DATE*      0
END_DATE*        0
CATEGORY*        0
START*           0
STOP*            0
MILES*           0
PURPOSE*       502
dtype: int64

In [14]:
uber.isnull().sum().sum()   # displaying the total number of missing values across all the variables

np.int64(502)

In [15]:
uber.describe()   # Summary of the original data including : Count,Mean, Std, Min, 25%,50%,75% and max

,MILES*
count,1155.000000
mean,10.566840
std,21.579106
min,0.500000
25%,2.900000
50%,6.000000
75%,10.400000
max,310.300000


In the above summary, we can see that the maximum miles driven by an uber drive is 310, and the minimum miles is 0.50



## Q9. Drop the missing values and store data in a new dataframe (name it"df") (2-points)

#### Note: Dataframe "df" will not contain any missing value

In [ ]:
# first copied the original dataset(uber_drives) to 'new_dataset'
uber_copy = uber_drives.copy()

# droped the missing values and stored the data in a new dataframe 'df'
df = uber_copy.dropna()

#display 'df' which does not contain any missing values
df.head()

## Q10. Check the information of the dataframe(df). (2 points)
#### Hint: Information includes - Total number of columns,variable data-types, number of non-null values in a variable, and usage

In [ ]:
df.info()    #dataframe 'df' information

In [ ]:
#renaming columns


df = df.rename(columns = {'START_DATE*':'start_date',\
                     'END_DATE*':'end_date',\
                     'CATEGORY*':'category',\
                     'START*':'start',\
                     'STOP*':'stop',\
                     'MILES*':'miles',\
                     'PURPOSE*':'purpose'})


In [ ]:
df.head()

## Q11. Get the unique start destinations. (2 points)
#### Note: This question is based on the dataframe with no 'NA' values
#### Hint- You need to print the unique destination place names in this and not the count.

In [ ]:
print(df['start'].unique())   # getting the unique START places names

In [ ]:
df['start'].nunique()

In [ ]:
df['start'].value_counts().head(10)

## Q12. What is the total number of unique start destinations? (2 points)
#### Note: Use the original dataframe without dropping 'NA' values

In [ ]:
# displaying total number of unique START destinations of original dataframe 'uber_drives'

uber_drives['START*'].nunique()

## Q13. What is the total number of unique stop destinations. (2 points)
#### Note: Use the original dataframe without dropping 'NA' values.

In [ ]:
# displaying total number of unique STOP destinations of original dataframe 'uber_drives'

uber_drives['STOP*'].nunique()

## Q14. Display all the Uber trips that has the starting point of San Francisco. (2 points)
#### Note: Use the original dataframe without dropping the 'NA' values.

#### Hint: You need to display the rows which has starting point of San Francisco. Try using loc function

In [ ]:
# First copied the orginal dataframe into new dataframe 'uber_drives_2'
uber_drives_2 = pd.DataFrame(uber_drives).copy()

# Now, select the index column as 'START'
uber_drives_2 = pd.read_csv(r'D:\My Data\Professional\Analytics\PGPDSBA-Projects\Proj.01 - Python for Data Analysis\Uber_Drive.csv', index_col='START*')

# selecting the rows which has starting point of San Francisco using loc function and assign it to new variable
uber_trips_SF = uber_drives_2.loc['San Francisco']


# displaying all the uber trips that has starting point of San Francisco
uber_trips_SF

In [ ]:
uber_trips_SF.value_counts()   # summing up all the numerical columns of the dataframe 'uber_trips_SF'

## Q15. What is the most popular starting point for the Uber drivers? (2 points)
#### Note: Use the original dataframe without dropping the 'NA' values.

#### Hint:Popular means the place that is visited the most

In [ ]:
# First copying the original dataframe into another dataframe named 'uber_drives_3'
uber_drives_3 = pd.DataFrame(uber_drives).copy()

print('The most popular starting point for the Uber drivers is:', uber_drives_3['START*'].value_counts().idxmax())
print('_____________________')
print('See the below data to verify:\n\n', uber_drives_3['START*'].value_counts())


## Q16. What is the most popular dropping point for the Uber drivers? (2 points)
#### Note: Use the original dataframe without dropping the 'NA' values.

#### Hint: Popular means the place that is visited the most

In [ ]:
# First copying the original dataframe into another dataframe named 'uber_drives_4'
uber_drives_4 = pd.DataFrame(uber_drives).copy()

print('The most popular dropping point for the Uber drivers is:', uber_drives_4['STOP*'].value_counts().idxmax())
print('_________________________')
print('See the below data to verify:\n\n', uber_drives_4['STOP*'].value_counts())


## Q17. List the most frequent route taken by Uber drivers. (3 points)
#### Note: This question is based on the new dataframe with no 'na' values.
#### Hint-Print the most frequent route taken by Uber drivers (Route= combination of START & END points present in the Data set). One may use Groupby function

In [ ]:
F_route = df.groupby(['start', 'stop']).sum()

F_route



## Q18. Display all types of purposes for the trip in an array. (3 points)
#### Note: This question is based on the new dataframe with no 'NA' values.

In [ ]:
#displaying all types of purposes for the trip in an array from the dataframe 'df' with no 'NA' values

ptype = df['purpose'].unique()

np.array(ptype)


## Q19. Plot a bar graph of Purpose vs Miles(Distance). (3 points)
#### Note: Use the original dataframe without dropping the 'NA' values.
#### Hint:You have to plot total/sum miles per purpose

In [ ]:
uber_drives.groupby(['PURPOSE*']).sum()

new_dataset = {'Airport/Travel':16.5, 'Between Offices':197.0, 'Charity($)':15.1, 'Commute':180.2, 'Customer Visit':2089.5,
              'Errand/Supplies':508.0, 'Meal/Entertain':911.7, 'Meeting':2851.3, 'Moving':18.2, 'Temporary Site':523.7}


purpose = list(new_dataset.keys())
miles = list(new_dataset.values())

fig = plt.figure(figsize = (17, 10))

plt.bar(purpose, miles, color ='green', width = 0.4)
plt.title("Total/Sum miles per purpose")
plt.xlabel("purpose")
plt.ylabel("miles")

plt.show()

## Q20. Display a dataframe of Purpose and the distance travelled for that particular Purpose. (3 points)
#### Note: Use the original dataframe without dropping "NA" values

In [ ]:
# Displaying a dataframe of Purpose and the distance travelled for that particular Purpose

uber_drives.groupby(['PURPOSE*']).sum()

## Q21. Plot number of trips vs Category of trips. (3 points)
#### Note: Use the original dataframe without dropping the 'NA' values.
#### Hint : You can make a countplot or barplot.

In [ ]:
uber_drives.groupby(['CATEGORY*']).groups



## Q22. What is proportion of miles that are covered as Business trips and what is the proportion of miles that are covered as Personal trips? (3 points)

### Note:Use the original dataframe without dropping the 'NA' values. The proportion calculation  is with respect to the 'miles' variable.

#### Hint: Proportion of miles covered as business trips= (Total Miles clocked as Business Trips)/ (Total Miles)
#### Proportion of miles covered as personal trips= (Total Miles clocked as Personal Trips)/ (Total Miles)


In [ ]:
total_miles = uber_drives['MILES*'].sum()    # sum of miles = 12204.7
print('Total Miles = ',total_miles)

print('______________________')

miles_per_category = uber_drives.groupby(['CATEGORY*']).sum()

print('\nMiles per category are shown as below = \n\n', miles_per_category)

print('______________________')

#Proportion of miles covered as business trips= (Total Miles clocked as Business Trips)/ (Total Miles)

Proportion_of_miles_covered_as_business_trips = (11487.0)/(12204.7)
print('Proportion of miles covered as business trips: ',Proportion_of_miles_covered_as_business_trips)

print('______________________')

#Proportion of miles covered as personal trips= (Total Miles clocked as Personal Trips)/ (Total Miles)

Proportion_of_miles_covered_as_personal_trips = (717.7)/(12204.7)
print('Proportion of miles covered as personal trips: ',Proportion_of_miles_covered_as_personal_trips)

